File: crates/pyxis-app/src/handlers/files.rs
use chrono::Utc;
use pyxis_db::{
    database::Database,
    entities::files::{Files, Link},
};
use tauri::State;

#[tauri::command]
pub fn create_file(
    title: String,
    dir_uid: Option<String>,
    path: String,
    workspace_uid: String,
    links: Vec<Link>,
    tags: Vec<String>,
    created_at: Option<String>,
    updated_at: Option<String>,
    uid: Option<String>,
    database: State<Database>,
) -> Option<Files> {
    let file = Files::new(dir_uid, path, title, tags, links, workspace_uid, None, created_at, updated_at, uid);

    match file.create(&database.get_connection()) {
        Ok(_) => Some(file),
        Err(e) => {
            eprintln!("[Files] Failed to create! {}", e);
            None
        }
    }
}

#[tauri::command]
pub fn list_files(
    workspace_uid: String,
    dir_uid: Option<String>,
    database: State<Database>,
) -> Option<Vec<Files>> {
    match Files::list(&database.get_connection(), workspace_uid, dir_uid) {
        Ok(directories) => Some(directories),
        Err(e) => {
            eprintln!("[Files] Failed to fetch! Error: {e}");
            None
        }
    }
}

#[tauri::command]
pub fn delete_file(uid: String, database: State<Database>) -> bool {
    match Files::delete(uid, &database.get_connection()) {
        Ok(_) => true,
        Err(e) => {
            eprintln!("[Files] Failed to delete! {}", e);
            false
        }
    }
}

#[tauri::command]
pub fn update_file(
    uid: String,
    title: String,
    dir_uid: Option<String>,
    workspace_uid: String,
    path: String,
    links: Vec<Link>,
    tags: Vec<String>,
    database: State<Database>,
) -> Option<Files> {
    let conn = &database.get_connection();
    let file = match Files::list(conn, workspace_uid, dir_uid.clone()) {
        Ok(f) => f.into_iter().find(|f| f.uid == uid),
        Err(e) => {
            eprintln!("[Files] Failed to get for update! {}", e);
            None
        }
    };

    if let Some(mut file) = file {
        file.title = title;
        file.dir_uid = dir_uid;
        file.path = path;
        file.links = links;
        file.tags = tags;
        file.updated_at = Utc::now().to_rfc3339();

        match file.update(conn) {
            Ok(_) => {
                return Some(file);
            }
            Err(e) => {
                eprintln!("[Files] Failed to update! {}", e);
                return None;
            }
        }
    }

    None
}

#[tauri::command]
pub fn get_file_id(path: String, workspace_uid: String, database: State<Database>) -> Option<i64> {
    match Files::get_by_path(&database.get_connection(), path, workspace_uid) {
        Ok(file_id) => Some(file_id),
        Err(e) => {
            eprintln!("[Files] Failed to fetch id! Error: {e}");
            None
        }
    }
}

File: packages/client/src/ffi/files.ts
import { invoke } from "./invoke";

export type Link = {
  title: string;
  url: string;
};

export type File = {
  id?: number;
  uid: string;
  dir_uid?: string;
  title: string;
  path: string;
  workspace_uid: string;
  created_at: string;
  updated_at: string;
  tags: Array<string>;
  links: Array<Link>;
  synced?: boolean;
};

type Args = {
  list_files: { dirUid?: string; workspaceUid: string };
  create_file: {
    title: string;
    workspaceUid: string;
    path: string;
    dirUid?: string;
    links: Array<Link>;
    tags: Array<string>;
    uid?: string;
    created_at?: string;
    updated_at?: string;
    synced?: boolean;
  };
  get_file_id: {
    path: string;
    workspaceUid: string;
  };
  delete_file: { uid: string };
  update_file: {
    uid: string;
    title: string;
    dirUid?: string;
    workspaceUid: string;
    path: string;
    links: Array<Link>;
    tags: Array<string>;
    synced?: boolean;
  };
};

export const createFile = async (
  title: string,
  workspaceUid: string,
  path: string,
  links: Array<Link>,
  tags: Array<string>,
  dirUid?: string,
  uid?: string,
  createdAt?: string,
  updatedAt?: string,
  synced?: boolean,
) => {
  try {
    const file = await invoke<Args, File | null>()("create_file", {
      title,
      path,
      workspaceUid,
      links,
      tags,
      dirUid,
      uid,
      created_at: createdAt,
      updated_at: updatedAt,
      synced,
    });

    if (!file) {
      return;
    }

    return file;
  } catch (e) {
    console.error("[File] Create failed!", e);
  }
};

export const getFiles = async (workspaceUid: string, dirUid?: string) => {
  try {
    const files = await invoke<Args, Array<File>>()("list_files", {
      workspaceUid,
      dirUid,
    });

    if (!files) {
      throw "Unknown failure!";
    }

    return files;
  } catch (e) {
    console.error("[File] Failed to fetch!", e);
    throw e;
  }
};

export const getFileByPath = async (path: string, workspaceUid: string) => {
  try {
    return await invoke<Args, number | undefined>()("get_file_id", {
      workspaceUid,
      path,
    });
  } catch (e) {
    console.error("[File] Failed to fetch by path!", e);
  }
};

export const deleteFile = async (uid: string) => {
  try {
    if (
      !(await invoke<Args, boolean>()("delete_file", {
        uid,
      }))
    ) {
      throw new Error("Empty Response!");
    }

    return;
  } catch (e) {
    console.error("[File] Failed to delete!", e);
    throw e;
  }
};

export const updateFile = async (
  uid: string,
  title: string,
  workspaceUid: string,
  path: string,
  links: Array<Link>,
  tags: Array<string>,
  dirUid?: string,
  synced?: boolean,
) => {
  try {
    const file = await invoke<Args, File | null>()("update_file", {
      uid,
      title,
      workspaceUid,
      path,
      dirUid,
      tags,
      links,
      synced,
    });

    if (!file) {
      throw new Error("Empty Response!");
    }

    return file;
  } catch (e) {
    console.error("[File] Failed to update!", e);
    throw e;
  }
};
